### PR TITLE
Drill 4236: ExternalSort should use the new allocator functionality to better manage it's memory usage

### PR DESCRIPTION
**Pull request by adeneche**
@StevenMPhillips can you please review ?

### CODE REVIEW

**Comment by StevenMPhillips** [comment]
+1 LGTM

**Comment by adeneche** [comment]
Thanks @StevenMPhillips , I will squash the commits, run one final round of tests and commit to master

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java [hunk 1]
```diff
@@ -588,36 +571,29 @@ public BatchGroup mergeAndSpill(LinkedList<BatchGroup> batchGroups) throws Schem
...
-        // Spilling this.batchGroups won't help here as they are owned by upstream operator,
-        // but spilling spilledBatchGroups may free enough memory
-        final BatchGroup merged = mergeAndSpill(spilledBatchGroups);
+        final BatchGroup merged = mergeAndSpill(batchGroups);
         if (merged != null) {
-          spilledBatchGroups.addFirst(merged);
+          spilledBatchGroups.add(merged);
         } else {
-          throw new OutOfMemoryException("Unable to allocate sv2, and not enough batchGroups to spill");
+          final String message = String.format(
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: Doesn't the Userexception.memoryError() already have this functionality?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java [hunk 2]
```diff
@@ -588,36 +571,29 @@ public BatchGroup mergeAndSpill(LinkedList<BatchGroup> batchGroups) throws Schem
...
     if (!sv2.allocateNewSafe(incoming.getRecordCount())) {
       try {
-        // Not being able to allocate sv2 means this operator's allocator reached it's maximum capacity.
-        // Spilling this.batchGroups won't help here as they are owned by upstream operator,
-        // but spilling spilledBatchGroups may free enough memory
-        final BatchGroup merged = mergeAndSpill(spilledBatchGroups);
+        final BatchGroup merged = mergeAndSpill(batchGroups);
         if (merged != null) {
-          spilledBatchGroups.addFirst(merged);
+          spilledBatchGroups.add(merged);
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: Out of curiosity, what is this change? It seems strange.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java (hunk 1 above)
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: what functionality ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java (hunk 2 above)
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: I'm not sure actually. We use `addFirst` in 2 places only and `add` in 3 other places. Maybe I should just change all of them to use `add` instead.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java (hunk 1 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: The string format construction.

File: common/src/main/java/org/apache/drill/common/exceptions/UserException.java [hunk 3]
```diff
@@ -50,7 +50,15 @@
    */
   public static Builder memoryError(final Throwable cause) {
     return UserException.resourceError(cause)
-      .message(MEMORY_ERROR_MSG);
+      .message(MEMORY_ERROR_MSG).addContext(cause.getMessage());
+  }
+
+  public static Builder memoryError(final String message) {
+    Builder builder =  UserException.resourceError();
+    if (!message.isEmpty()) {
```
Comment (adeneche) [inline common/src/main/java/org/apache/drill/common/exceptions/UserException.java]: Oups, this method doesn't set the proper error message to the user exception, it's causing TestOutOfMemoryOutcome to fail.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java (hunk 2 above)
Comment (StevenMPhillips) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: The ordering in the queue only matters for determining which BatchGroups will be merged the next time mergeAndSpill is called. Since mergeAndSpill takes batches off the the end of the queue, we want to put the outcome at the front of the queue, since the outcome BatchGroup has a much larger spill size than the ones at the end of the queue, and is thus more expensive to merge.

So I think you should revert this change, and leave it the way it is. It might be a good idea to take more extensive look at the spilling/merging strategy, but that can be a different task.

Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: One possible explanation, and @StevenMPhillips could confirm it:

sort tries to prevent `spilledBatchGroups` from becoming too large by spilling batches from this group regularly to disk, by putting the result of those spills at the beginning of spilledBatchGroups  we avoid spilling the same batches over and over again

Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: will revert the change.

I'm confused though, why do we sometimes (in 3 different places) add the outcome of `mergeAndSpill` at the end of the queue ?

Comment (StevenMPhillips) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: It depends on which queue we are merging from. When batches first come in to the ExternalSort, they are added to the batchGroup queue. Usually when spilling, we spill some number of batches from the batchGroup queue, and then put the result at the end of the spilledBatchGroup queue. In the cases where we need to re-merge the batchGroups, we take from the end of the spilledBatchGroup queue, and add them to the front of the queue.

If you are seeing a place where we mergeAndSpill the spilledBatchGroup queue and add to the end of the queue, that is a mistake and should be fixed.

Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: thanks Steven, I'm less confused now! The change is valid, because we are merging from the batchGroups queue.

Comment (StevenMPhillips) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: Yeah, since you also changed it so that we are spilling the batchGroup queue, this change does make sense.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java (hunk 1 above)
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java]: now it does
